### PR TITLE
Gas Cost Checks for Precompile Calls

### PR DESCRIPTION
**Pull request by roynalnaruto**
### Description

The PR constrains the `step_gas_cost` in `CALLOP` specifically in the case of precompile calls.

### Progress

- [x] Ecrecover
- [x] EcAdd
- [x] EcMul
- [x] EcPairing
- [x] Modexp
- [x] Identity

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/callop.rs [hunk 1]
```diff
@@ -296,6 +296,10 @@ impl<F: Field> ExecutionGadget<F> for CallOpGadget<F> {
                     no_callee_code.expr(),
                     true.expr(),
                 );
+                cb.require_true(
+                    "Precompile addresses are always warm",
+                    and::expr([is_warm.expr(), is_warm_prev.expr()]),
+                );
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs:310]: in begin tx we already set is_warm of precompiles to true, do you think we still need check here?

https://github.com/scroll-tech/zkevm-circuits/blob/d0e7a07e8af25220623564ef1c3ed101ce63220e/zkevm-circuits/src/evm_circuit/execution/begin_tx.rs#L252-L255

File: zkevm-circuits/src/evm_circuit/execution/precompiles/modexp.rs [hunk 2]
```diff
@@ -1032,7 +1195,7 @@ mod test {
                         PUSH32(word!("0x0000000000000000000000000000000000000000000000000000000000000008"))
                         PUSH1(0x60)
                         MSTORE
-                        PUSH32(word!("0x0000000000000000000000000000000000000000000000000000000000000009"))
+                        PUSH32(word!("0x1000000000000000000000000000000000000000000000000000000000000009"))
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/modexp.rs:1198]: this line is used to test modexp gas for large input?

**Review by lispc — APPROVED** [review]

**Comment by lispc** [comment]
@noel2004 can also have a glance

File: zkevm-circuits/src/evm_circuit/execution/precompiles/modexp.rs (hunk 2 above)
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/precompiles/modexp.rs:1198]: There were no test cases where `exponent`'s bit length was significant enough for the gas cost to be > 200. So this line changes the exponent's bit length to a high value, hence testing the gas calculation gadget.

File: zkevm-circuits/src/evm_circuit/execution/precompiles/ec_add.rs [hunk 3]
```diff
@@ -57,6 +59,12 @@ impl<F: Field> ExecutionGadget<F> for EcAddGadget<F> {
             cb.query_cell_phase2(),
             cb.query_cell_phase2(),
         );
+        let gas_cost = cb.query_cell();
```
Comment (naure) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/ec_add.rs:62]: There is no need for a cell here. The constant expression could be used instead.

**Pull request merged**